library/psa_crypto.c

```diff
@@ -3131,7 +3135,690 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+    operation->num_ops = 0;
+
+    status = psa_driver_wrapper_sign_hash_start( operation, &attributes,
+                                                 slot->key.data,
+                                                 slot->key.bytes, alg,
+                                                 hash, hash_length );
+exit:
+
+    if( status != PSA_SUCCESS )
+        psa_sign_hash_abort_internal( operation );
+
+    unlock_status = psa_unlock_key_slot( slot );
+
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+
+}
+
+
+psa_status_t psa_sign_hash_complete(
+                            psa_sign_hash_interruptible_operation_t *operation,
+                            uint8_t *signature, size_t signature_size,
+                            size_t *signature_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    *signature_length = 0;
+
+    /* Check that start has been called first. */
+    if( operation->id == 0 )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    /* Immediately reject a zero-length signature buffer. This guarantees
+    * that signature must be a valid pointer. (On the other hand, the input
+    * buffer can in principle be empty since it doesn't actually have
+    * to be a hash.) */
+    if( signature_size == 0 )
+    {
+        status = PSA_ERROR_BUFFER_TOO_SMALL;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_sign_hash_complete( operation, signature,
+                                                    signature_size,
+                                                    signature_length );
+exit:
+
+    /* Update ops count with work done. */
+    operation->num_ops += psa_driver_wrapper_sign_hash_get_num_ops( operation );
+
+    if( status != PSA_OPERATION_INCOMPLETE )
+    {
+        /* Fill the unused part of the output buffer (the whole buffer on error,
+         * the trailing part on success) with something that isn't a valid
+         * signature (barring an attack on the signature and
+         * deliberately-crafted input), in case the caller doesn't check the
+         * return status properly.*/
+        if( status == PSA_SUCCESS )
+            memset( signature + *signature_length, '!',
+                    signature_size - *signature_length );
+        else if( signature_size > 0 )
+            memset( signature, '!', signature_size );
+            /* If signature_size is 0 then we have nothing to do. We must not
+             * call memset because signature may be NULL in this case.*/
+
+        psa_sign_hash_abort_internal( operation );
+    }
+
+    return( status );
+}
+
+psa_status_t psa_sign_hash_abort(
+                            psa_sign_hash_interruptible_operation_t *operation )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    status = psa_sign_hash_abort_internal( operation );
+
+    /* We clear the number of ops done here, so that it is not cleared when
```

Comment:
I am unsure what you mean. The global this clears is at the PSA scope (and will be even more so once I remove the driver entry points for setting and getting it), so surely it should be cleared here, and the comment with it?

Admittedly we could memset, but this would clear the contained contexts, although they should have admittedly already been cleared by this point. Is this acceptable?

Reply:
Typically aborting an operation means: 1. free auxiliary resources (such as heap-allocated blocks for which there is a pointer in the context); 2. memset the context to 0. And that's what we normally do when an operation reaches an error condition or is finished. In the case of restartable operations, `complete()` finishes the operation but there is one field in the context that we don't zero out (because we want `get_ops` to keep returning information about the aborted/completed operation). This is unusual, so we should explain in a comment in the internal abort function.